### PR TITLE
Validate hotkey when setting it

### LINKED ISSUE #1755: flow launcher got an error
[2023-01-01.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/10328438/2023-01-01.txt)

Exception msg:

Flow Launcher version: 1.11.0
OS Version: Microsoft Windows NT 10.0.19044.0
IntPtr Length: 8
x64: True

Python Path: C:\Users\Rishu Mishra.DESKTOP-7S0PFJC\AppData\Roaming\FlowLauncher\Environments\Python\PythonEmbeddable-v3.8.9\pythonw.exe
Node Path: C:\Program Files\nodejs\node.exe
Date: 01/01/2023 12:16:17
Exception:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Flow.Launcher.MainWindow.VerticalCenter() in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 540
   at Flow.Launcher.MainWindow.InitializePosition() in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 212
   at Flow.Launcher.MainWindow.UpdatePosition() in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 499
   at Flow.Launcher.MainWindow.HideStartup() in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 515
   at Flow.Launcher.MainWindow.OnLoaded(Object sender, RoutedEventArgs _) in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 93
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
   at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
   at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.Resize(ICompositionTarget resizedCompositionTarget)
   at System.Windows.Interop.HwndTarget.OnResize()
   at System.Windows.Interop.HwndTarget.HandleMessage(WindowMessage msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
- Prevent user from setting single "Alt" "Ctrl" "Shift" "Win" as hotkey. Fix #1755.
- Prevent setting single alphanumeric key as hotkey.
- - When a pressed hotkey is unavailable, stay focused inside the text box so user can continue to try a new hotkey.
- When clicked away and the hotkey is unavailable, return to the previous set hotkey.

Tests:
- Toggle hotkey can be modified.
- Preview hotkey can be modified.
- Single special key can't be set as hotkey.
- Single function key can be set as hotkey.
- Single alphanumeric key can't be set as hotkey.
- When a pressed hotkey is unavailable, stay focused inside the text box so user can continue to try a new hotkey.
- When clicked away and the hotkey is unavailable, return to the previous set hotkey.

### CODE REVIEW

File: Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs [hunk 1]
```diff
@@ -86,7 +97,7 @@ private void Parse(string hotkeyString)
                 Ctrl = true;
                 keys.Remove("Ctrl");
             }
-            if (keys.Count > 0)
+            if (keys.Count == 1)
```
Comment (taooceros) [inline Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs:100]: Will user want to use multiple key as a combo?

Comment (VictoriousRaptor) [reply Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs:100]: > Will user want to use multiple key as a combo?

It's not supported. Haven't seen any software supports "ctrl+A+D"

Comment (taooceros) [reply Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs:100]: Oh we don't support it? I thought this is supported before sorry

**Review by jjw24 — CHANGES_REQUESTED** [review]
Can we:
1. When a hotkey is unavailable, stay focused inside the text box so user can continue to try new hotkey.
2. When clicked away and the hotkey is unavailable, return to the previous set hotkey.

**Comment by taooceros** [comment]
I just remember I have a very old pr regarding this #933.

**Comment by VictoriousRaptor** [comment]
Closed by accident somehow.

**Comment by VictoriousRaptor** [comment]
> Can we:
> 
>     1. When a hotkey is unavailable, stay focused inside the text box so user can continue to try new hotkey.
> 
>     2. When clicked away and the hotkey is unavailable, return to the previous set hotkey.

Please review again.

File: Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs [hunk 2]
```diff
@@ -110,36 +111,75 @@ private void Parse(string hotkeyString)
...
+                case Key.RightShift:
+                case Key.LWin:
+                case Key.RWin:
+                    return false;
+                default:
+                    if (ModifierKeys == ModifierKeys.None)
+                    {
+                        return !((CharKey >= Key.A && CharKey <= Key.Z) ||
+                            (CharKey >= Key.D0 && CharKey <= Key.D9) ||
+                            (CharKey >= Key.NumPad0 && CharKey <= Key.NumPad9));
```
Comment (VictoriousRaptor) [inline Flow.Launcher.Infrastructure/Hotkey/HotkeyModel.cs:159]: Checking for all symbol keys are too difficult so I just ban alphanumeric keys.

**Review by jjw24 — APPROVED** [review]
👍

**Pull request merged**